contracts/registrar/contract/registrar.sol

```diff
@@ -0,0 +1,214 @@
+    event NewCheckpointEvent(uint indexed index, address grantor, bytes32 checkpointHash);
+
+    // AddAdminEvent is emitted when new address is accepted as admin.
+    // Grantor indicates who authorizes the add admin operation.
+    event AddAdminEvent(address addr, address grantor, string description);
+
+    // RemoveAdminEvent is emitted when an admin is removed.
+    // Grantor indicates who authorizes the remove admin operation.
+    event RemoveAdminEvent(address addr, address grantor, string reason);
+
+    /*
+        Public Functions
+    */
+    constructor(address[] _adminlist) public {
+        // regard contract creator as a default admin.
+        admins[msg.sender] = 1;
+        adminList.push(msg.sender);
+        for (uint i = 0; i < _adminlist.length; i++) {
+            admins[_adminlist[i]] = 1;
+            adminList.push(_adminlist[i]);
+        }
+    }
+
+    /**
+     * @dev Get latest stable checkpoint information.
+     * @return section index
+     * @return checkpoint hash
+     */
+    function GetLatestCheckpoint()
+    view
+    public
+    returns(uint, bytes32) {
+        bytes32 hash = GetCheckpoint(latest);
+        return (latest, hash);
+    }
+
+    /**
+     * @dev Get a stable checkpoint information with specified section index.
+     * @param _sectionIndex section index
+     * @return checkpoint hash
+     */
+    function GetCheckpoint(uint _sectionIndex)
+    view
+    public
+    returns(bytes32)
+    {
+        return checkpoints[_sectionIndex];
+    }
+
+    /**
+     * @dev Set stable checkpoint information.
+     * Checkpoint represents a set of post-processed trie roots (CHT and BloomTrie)
+     * associated with the appropriate section head hash.
+     *
+     * It is used to start light syncing from this checkpoint
+     * and avoid downloading the entire header chain while still being able to securely
+     * access old headers/logs.
+     *
+     * Note we trust the given information here provided by foundation,
+     * need a trust less version for future.
+     * @param _sectionIndex section index
+     * @param _sectionHead section header
+     * @param _chtRoot cht root hash
+     * @param _bloomTrieRoot bloom trie root hash
+     * @return indicator whether set checkpoint successfully
+     */
+    function SetCheckpoint(
+        uint _sectionIndex,
+        bytes32 _sectionHead,
+        bytes32 _chtRoot,
+        bytes32 _bloomTrieRoot
+    )
+    OnlyAuthorized
+    public
+    returns(bool)
+    {
+        // Ensure the checkpoint information provided is strictly continuous with previous one.
+        // But the latest checkpoint modification is allowed.
+        if (_sectionIndex != latest && _sectionIndex != latest + 1 && latest != 0) {
+            return false;
+        }
+        // Ensure the checkpoint is stable enough to be registered.
+        if (block.number < (_sectionIndex+1)*sectionSize+confirmations) {
```

Comment:
Should we loose the restrict here? If not, it means external admin can only register new checkpoint in the range [(_sectionIndex+1)*sectionSize+256, (_sectionIndex+1)*sectionSize+8192], if admin forget to do something, it will be trouble.

Reply:
No, it applies only to modifying the already existing last one (_sectionIndex == latest). If no checkpoint has been registered for that section then it can be registered later too (in which case it will be instantly final). Not a perfect rule but we assume trusted admins here anyway and in normal operation the checkpoints will be registered soon after +256 blocks so there is time to detect if there is a disagreement about the checkpoint. The imporant part here is that there is a way to correct errors and still clients can be sure about finality.